### ISSUE TITLE
Unable to unbox nested arrays

### ISSUE THREAD

**Issue by bencallis** [JohnSundell/Unbox#84]
I have been using Unbox and I am really enjoying the simplicity of it. Today I have hit a strange issue while trying to unbox the following structure

```
{
  .... ,
  "legs": [
    [
      {
        "cabin_class": null,
        "from_airport_name": "Paris Charles-de-Gaulle",
        "marketing_name": null,
        "to_airport_name": "Londres Heathrow",
        "to_airport_iata": "LHR",
        "marketing_id": "BA",
        "operator_name": null,
        "arr_datetime": "2016-09-16T07:45:00.000000",
        "dep_datetime": "2016-09-16T07:25:00.000000",
        "operator_id": null,
        "from_airport_iata": "CDG",
        "flight_num": "0303"
      },
      {
        "cabin_class": null,
        "from_airport_name": "Londres Heathrow",
        "marketing_name": null,
        "to_airport_name": "New York John F Kennedy",
        "to_airport_iata": "JFK",
        "marketing_id": "BA",
        "operator_name": null,
        "arr_datetime": "2016-09-16T12:10:00.000000",
        "dep_datetime": "2016-09-16T09:40:00.000000",
        "operator_id": null,
        "from_airport_iata": "LHR",
        "flight_num": "0175"
      }
    ],
    [
      {
        "cabin_class": null,
        "from_airport_name": "New York John F Kennedy",
        "marketing_name": null,
        "to_airport_name": "Londres Heathrow",
        "to_airport_iata": "LHR",
        "marketing_id": "BA",
        "operator_name": null,
        "arr_datetime": "2016-09-20T07:45:00.000000",
        "dep_datetime": "2016-09-19T19:30:00.000000",
        "operator_id": null,
        "from_airport_iata": "JFK",
        "flight_num": "0176"
      },
      {
        "cabin_class": null,
        "from_airport_name": "Londres Heathrow",
        "marketing_name": null,
        "to_airport_name": "Paris Charles-de-Gaulle",
        "to_airport_iata": "CDG",
        "marketing_id": "BA",
        "operator_name": null,
        "arr_datetime": "2016-09-20T11:15:00.000000",
        "dep_datetime": "2016-09-20T08:50:00.000000",
        "operator_id": null,
        "from_airport_iata": "LHR",
        "flight_num": "0306"
      }
    ]
  ]
}
```

```
class BookingSummary: Unboxable {

    .....
    let legs: [[Leg]] ** array of  arrays containing legs

    
    private struct JSONKey {
        ....
        static let legs = "legs"
    }
    
    required init(unboxer: Unboxer) {
        ....
        self.legs = unboxer.unbox(JSONKey.legs) **** Fails here
    }
}
```

This results in an Unbox error when unboxing legs. Accessing the first (or any other valid index) element using key path works 
```
    let legs: [Leg]

//
 self.legs = unboxer.unbox("legs.0", isKeyPath: true)
```

Should nested arrays work? Or do I need to make an intermediary object that just contains the legs array?

Leg class

```
import Unbox

class Leg: Unboxable {

    let cabinClass: String?
    let fromAirportName: String
    let marketingName: String?
    let toAirportName: String
    let toAirportIata: String
    let marketingId: String
    let operatorName: String?
    let arrDatetime: String
    let depDatetime: String
    let flightNum: String
    let operatorId: String?
    let fromAirportIata: String
    ...

    private struct JSONKey {
        static let cabinClass = "cabin_class"
        static let fromAirportName = "from_airport_name"
        static let marketingName = "marketing_name"
        static let toAirportName = "to_airport_name"
        static let toAirportIata = "to_airport_iata"
        static let marketingId = "marketing_id"
        static let operatorName = "operator_name"
        static let arrDatetime = "arr_datetime"
        static let depDatetime = "dep_datetime"
        static let operatorId = "operator_id"
        static let fromAirportIata = "from_airport_iata"
        static let flightNum = "flight_num"
        ...
    }
    
    required init(unboxer: Unboxer) {
        self.toAirportIata = unboxer.unbox(JSONKey.toAirportIata)
        self.toAirportName = unboxer.unbox(JSONKey.toAirportName)
        self.operatorName = unboxer.unbox(JSONKey.operatorName)
        self.marketingId = unboxer.unbox(JSONKey.marketingId)
        self.fromAirportName = unboxer.unbox(JSONKey.fromAirportName)
        self.depDatetime = unboxer.unbox(JSONKey.depDatetime)
        self.arrDatetime = unboxer.unbox(JSONKey.arrDatetime)
        self.flightNum = unboxer.unbox(JSONKey.flightNum)
        self.cabinClass = unboxer.unbox(JSONKey.cabinClass)
        self.marketingName = unboxer.unbox(JSONKey.marketingName)
        self.operatorId = unboxer.unbox(JSONKey.operatorId)
        self.fromAirportIata = unboxer.unbox(JSONKey.fromAirportIata)
       ...
    }

}
```

**Comment by JohnSundell** [comment]
Hey @bencallis, nested arrays do indeed work, so since the first element is unboxable, I suspect that a subsequent member of the `legs` array contains invalid data. What happens if you `catch` the error that Unbox throws? Also could you post the code for your `Legs` struct, so I can check what might be wrong there? :)

**Comment by bencallis** [comment]
@JohnSundell Thanks for the incredibly quick reply! I have added the Leg class above. 
![unbox_swift](https://cloud.githubusercontent.com/assets/1911951/16518812/de1c7d52-3f7c-11e6-92ae-76404a8d1df3.png)

```
Printing description of invalidValue:
▿ Optional<protocol<>>
  ▿ Some : 2 elements
    ▿ [0] : 2 elements
      ▿ [0] : 13 elements
        ▿ [0] : 2 elements
          - .0 : operator_id
        ▿ [1] : 2 elements
          - .0 : marketing_id
          - .1 : BA
        ▿ [2] : 2 elements
          - .0 : to_airport_name
          - .1 : Londres Heathrow
        ▿ [3] : 2 elements
          - .0 : cabin_class
        ▿ [4] : 2 elements
          - .0 : from_airport_iata
          - .1 : CDG
        ▿ [5] : 2 elements
          - .0 : to_airport_iata
          - .1 : LHR
        ▿ [6] : 2 elements
          - .0 : skyscanner_code
          - .1 : BA { ... }
        ▿ [7] : 2 elements
          - .0 : marketing_name
        ▿ [8] : 2 elements
          - .0 : flight_num
          - .1 : 0303
        ▿ [9] : 2 elements
          - .0 : arr_datetime
          - .1 : 2016-09-16T07:45:00.000000
        ▿ [10] : 2 elements
          - .0 : operator_name
        ▿ [11] : 2 elements
          - .0 : dep_datetime
          - .1 : 2016-09-16T07:25:00.000000
        ▿ [12] : 2 elements
          - .0 : from_airport_name
          - .1 : Paris Charles-de-Gaulle
      ▿ [1] : 13 elements
        ▿ [0] : 2 elements
          - .0 : operator_id
        ▿ [1] : 2 elements
          - .0 : marketing_id { ... }
          - .1 : BA { ... }
        ▿ [2] : 2 elements
          - .0 : to_airport_name { ... }
          - .1 : New York John F Kennedy
        ▿ [3] : 2 elements
          - .0 : cabin_class { ... }
        ▿ [4] : 2 elements
          - .0 : from_airport_iata { ... }
          - .1 : LHR { ... }
        ▿ [5] : 2 elements
          - .0 : to_airport_iata { ... }
          - .1 : JFK
        ▿ [6] : 2 elements
          - .0 : skyscanner_code { ... }
          - .1 : BA { ... }
        ▿ [7] : 2 elements
          - .0 : marketing_name { ... }
        ▿ [8] : 2 elements
          - .0 : flight_num { ... }
          - .1 : 0175
        ▿ [9] : 2 elements
          - .0 : arr_datetime { ... }
          - .1 : 2016-09-16T12:10:00.000000
        ▿ [10] : 2 elements
          - .0 : operator_name { ... }
        ▿ [11] : 2 elements
          - .0 : dep_datetime { ... }
          - .1 : 2016-09-16T09:40:00.000000
        ▿ [12] : 2 elements
          - .0 : from_airport_name { ... }
          - .1 : Londres Heathrow { ... }
    ▿ [1] : 2 elements
      ▿ [0] : 13 elements
        ▿ [0] : 2 elements
          - .0 : operator_id
        ▿ [1] : 2 elements
          - .0 : marketing_id { ... }
          - .1 : BA { ... }
        ▿ [2] : 2 elements
          - .0 : to_airport_name { ... }
          - .1 : Londres Heathrow { ... }
        ▿ [3] : 2 elements
          - .0 : cabin_class { ... }
        ▿ [4] : 2 elements
          - .0 : from_airport_iata { ... }
          - .1 : JFK { ... }
        ▿ [5] : 2 elements
          - .0 : to_airport_iata { ... }
          - .1 : LHR { ... }
        ▿ [6] : 2 elements
          - .0 : skyscanner_code { ... }
          - .1 : BA { ... }
        ▿ [7] : 2 elements
          - .0 : marketing_name
        ▿ [8] : 2 elements
          - .0 : flight_num { ... }
          - .1 : 0176
        ▿ [9] : 2 elements
          - .0 : arr_datetime { ... }
          - .1 : 2016-09-20T07:45:00.000000
        ▿ [10] : 2 elements
          - .0 : operator_name { ... }
        ▿ [11] : 2 elements
          - .0 : dep_datetime { ... }
          - .1 : 2016-09-19T19:30:00.000000
        ▿ [12] : 2 elements
          - .0 : from_airport_name { ... }
          - .1 : New York John F Kennedy
      ▿ [1] : 13 elements
        ▿ [0] : 2 elements
          - .0 : operator_id { ... }
        ▿ [1] : 2 elements
          - .0 : marketing_id { ... }
          - .1 : BA { ... }
        ▿ [2] : 2 elements
          - .0 : to_airport_name { ... }
          - .1 : Paris Charles-de-Gaulle { ... }
        ▿ [3] : 2 elements
          - .0 : cabin_class { ... }
        ▿ [4] : 2 elements
          - .0 : from_airport_iata { ... }
          - .1 : LHR { ... }
        ▿ [5] : 2 elements
          - .0 : to_airport_iata { ... }
          - .1 : CDG { ... }
        ▿ [6] : 2 elements
          - .0 : skyscanner_code { ... }
          - .1 : BA { ... }
        ▿ [7] : 2 elements
          - .0 : marketing_name { ... }
        ▿ [8] : 2 elements
          - .0 : flight_num { ... }
          - .1 : 0306
        ▿ [9] : 2 elements
          - .0 : arr_datetime { ... }
          - .1 : 2016-09-20T11:15:00.000000
        ▿ [10] : 2 elements
          - .0 : operator_name { ... }
        ▿ [11] : 2 elements
          - .0 : dep_datetime { ... }
          - .1 : 2016-09-20T08:50:00.000000
        ▿ [12] : 2 elements
          - .0 : from_airport_name { ... }
          - .1 : Londres Heathrow { ... }
```

**Comment by bencallis** [comment]
Let me know if you need any further details :)

**Comment by JohnSundell** [comment]
@bencallis Will do! I'm out traveling so have very limited time & internet access, but will try to take a look tomorrow :)

**Comment by bencallis** [comment]
@JohnSundell Now rush! Enjoy your travels :)

**Comment by zwebie** [comment]
I'm having the same issue both with a nested array and an enum.
Although in my case I get a Xcode crash.
It worked fine in a previous version.

**Comment by clayellis** [comment]
@bencallis Are you still having issues with this? If so, I'll take some time today to look at it a little more closely.

**Comment by bencallis** [comment]
@clayellis Yeah I am still having this issue.

Current work around is

```
       // TODO: should be able to get all of the array response from Github issue https://github.com/JohnSundell/Unbox/issues/84
        let outbound: [Segment] = unboxer.unbox("legs.0", isKeyPath: true)
        if (legsJSON?.count < 2) {
            self.legs = [outbound]
        } else {
            let inbound: [Segment] = unboxer.unbox("legs.1", isKeyPath: true)
            self.legs = [outbound, inbound]
        }
```

**Comment by bencallis** [comment]
@clayellis did you get anywhere with this? May try and have a look into it next week. Any pointers would be handy.

**Comment by mjholgate** [comment]
I made an extension to workaround this issue:

```
extension Unboxer {

    // Workaround Github issue https://github.com/JohnSundell/Unbox/issues/84
    public static func unboxArrayOfArrayOfObject<T:Unboxable>(unboxer: Unboxer, key: String) -> [[T]] {
        let value = unboxer.dictionary[key]

        guard let outerArray = value as? [[UnboxableDictionary]] else {
            unboxer.failForInvalidValue(value, forKey: key)
            return []
        }

        do {
            return try outerArray.map { try $0.map { try Unbox($0) }}
        }
        catch {
            unboxer.failForInvalidValue(value, forKey: key)
            return []
        }
    }

}
```

**Comment by JohnSundell** [comment]
This has been fixed in Unbox 2.0! 🎉
